Network policies for the host endpoint

pchaigno opened the pull request:
This pull request adds network policies (CIDR, labels, and ports) for the host through the new host endpoint. It also introduces a new `nodeSelector` in our internal JSON policy.

As a summary:
1. Introduce a new `--enable-host-firewall` option.
0. Prepare BPF policy functions to accept a `localID` as argument.
0. Introduce host network policies on Go side.
0. Enforce IPv4 policies for the host on BPF side.
0. Enforce IPv6 policies for the host on BPF side.
0. Support egress LB.
0. Bypass host policies on paths from/to proxy.
0. Reduce BPF verifier complexity using `relax_verifier()` to improve state pruning.
0. Disable egress LB on older kernels if host policies are enabled to reduce complexity. 

The last piece, to watch and update labels for the node, will come in a separate pull request. It's also dependent on the host endpoint PR, but is independent from the present PR.

Fixes #9915

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -46,12 +46,23 @@ func (r Rule) Sanitize() error {
-	if r.EndpointSelector.LabelSelector == nil {
-		return fmt.Errorf("rule cannot have nil EndpointSelector")
+	if r.EndpointSelector.LabelSelector == nil && r.NodeSelector.LabelSelector == nil {
+		return fmt.Errorf("rule cannot have both nil EndpointSelector and nil NodeSelector")
```
nit, perhaps...?

```suggestion
		return fmt.Errorf("rule must have one of EndpointSelector or NodeSelector")
```

joestringer commented on pkg/policy/repository.go:
```diff
@@ -522,6 +522,10 @@ func (p *Repository) getMatchingRules(securityIdentity *identity.Identity) (ingr
 	egressMatch = false
 	for _, r := range p.rules {
+		isNode := securityIdentity.ID == identity.ReservedIdentityHost
+		if (r.NodeSelector.LabelSelector != nil) != isNode {
```
nit, might be slightly easier to follow like this?
```suggestion
		isNode := securityIdentity.ID == identity.ReservedIdentityHost
		selectsNode := r.NodeSelector.NodeSelector != nil
		if selectsNode != isNode {
```

joestringer commented on bpf/bpf_host.c:
```diff
@@ -760,34 +976,79 @@ int from_host(struct __ctx_buff *ctx)
+		// to-netdev is attached to the egress path of the native
+		// device.
+		srcID = ipcache_lookup_srcid4(ctx);
+		ret = ipv4_host_policy_egress(ctx, srcID, &remoteID);
```
How do you see the interaction between BPF_HOST_FIREWALL and non-BPF NAT including nodeport?

I'm thinking that if there are *any* SNAT or masquerade rules applying within the host, including for instance iptables-based kube-proxy NodePort implementation, that traffic will be subject to the host firewall, even if it was originally sourced from outside the node, because the SNAT is being applied before this and the ipcache lookup for the source will resolve to the host identity.

Given that this is an initial framework of the policy engine, it's probably not required to address for acceptance of this PR, but I'd be interested to discuss.

pchaigno replied:
:thinking: We would need some sort of mark to tell the host firewall to let those packets through. Don't we already have similar mechanisms to carry info. through non-BPF NATing?

joestringer replied:
Right, so for reference the way we do this today is basically we have an OUTPUT chain rule that explicitly marks the traffic as "from the host". OUTPUT chain implies that the socket is in the local network namespace so the process *must* be local. We can then treat all traffic without this mark as "remote" in some sense. If traffic from the outside world arrives to the node and is SNAT'd, then it will not hit this rule. When we handle this traffic, even if we look up the IPCache and it maps the IP to host identity, we will explicitly ignore the ipcache and treat the traffic as from the world instead.

One tricky part here is that we are not in control of the component which installs the SNAT iptables rules here, so we can't just modify those rules to provide this context.

pchaigno replied:
Ah, right, makes sense. Are we able to carry a mark through kube-proxy's SNATing though or does that risk being removed?

Maybe that's something to discuss at the next sig-datapath meeting :thinking:

joestringer replied:
Passing a mark through iptables rules established by other agents like kube-proxy can potentially be done but we don't have a lot of control here. A newer release of the other agent could always modify the way they deploy iptables rules in a way that breaks our use of the mark. This is part of the motivation for the way we do the marking currently.

But yeah this is a bigger topic than this PR and is one of the last remaining items for eliminating Cilium's reliance on iptables, good candidate for sig-datapth as it could get fairly deep.

joestringer replied:
(I might add; the mark bits we *do* use are selected because they don't overlap with the mark bits used by kube-proxy, so yes we can and do carry marks through kube-proxy's SNAT rules but there's definitely details to work through there)

joestringer commented on bpf/bpf_host.c:
```diff
@@ -282,6 +290,28 @@ int tail_handle_ipv6_from_netdev(struct __ctx_buff *ctx)
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	info = ipcache_lookup4(&IPCACHE_MAP, ip4->saddr, V4_CACHE_KEY_LEN);
```
Should this be `lookup_ip4_remote_endpoint()` to pick up hashmap-based LPM trie lookup on older kernels?

joestringer replied:
(Note: This has the potential to further increase complexity on older kernels, so you may want to consider following up separately on how to handle this case)

pchaigno replied:
Fixed. Hints for the verifier added in a later commit should give us some healthy margin here. I also checked locally with the 4.9 dev. VM.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -333,14 +363,178 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+	tuple.saddr = ip4->saddr;
+	l4_off = l3_off + ipv4_hdrlen(ip4);
+	ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_EGRESS,
+			 &ct_state, &monitor);
```
`monitor` here is used for aggregating packet events from the datapath, would be awesome if we could propagate this through to the monitor trace event logic to prevent generating events for every packet using this mechanism.

That said, I presume there's no regression through here as we would already generate monitor messages for all such traffic so this would be reasonable as a follow-up optimization. (It also has the potential to affect complexity, which I'd rather not tie this PR down with).

pchaigno replied:
Hm, it may require some code duplication and/or moving some calls to `send_trace_notify()` because currently, in some cases, we call `send_trace_notify()` in one tail call program and compute `monitor` in another.

I've made a note to look into it as a follow up.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -333,14 +363,178 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+			return ret;
+		break;
+
+	case CT_ESTABLISHED:
```
Did you compare this with the egress bpf_lxc path to determine whether that code should be copied here too?

Specifically:
```
        case CT_ESTABLISHED:
                /* Did we end up at a stale non-service entry? Recreate if so. */
                if (unlikely(ct_state.rev_nat_index != ct_state_new.rev_nat_index)) {
                        goto ct_recreate4;
                }
                break;
```

joestringer replied:
(I'm not overly familiar with the background here, just noticed some differences.)

pchaigno replied:
For future reference, ce43de13e2e093cede0839d263e84f32b0c890ed provides background into this bit of code and it's definitely needed.

EDIT: No wonder that background commit seemed familiar. I've already read it before and this piece of code for `CT_ESTABLISHED` is added in the subsequent commit that adds support for egress LB (`bpf: Support LB on host egress`) :relieved:

joestringer commented on bpf/bpf_host.c:
```diff
@@ -333,14 +363,178 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
 {
 	struct remote_endpoint_info *info = NULL;
 	struct ipv4_ct_tuple tuple = {};
+	__u32 __maybe_unused remoteID = WORLD_ID;
```
I note that this is declared on the stack here and a pointer passed into the policy functions, but I don't understand why it's necessary. Did you intend to pass it on somewhere later on?

(At a glance this seemed to apply to all similar cases of passing the remoteID pointer into the policy functions)

pchaigno replied:
I was able to remove that argument from the `ipv{4,6}_host_policy_egress()` functions, but it's still needed for the `ipv{4,6}_host_policy_ingress()` functions because there's one case (in `to_host()`) where we pass that ID into `send_drop_notify_error()`.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -716,8 +921,12 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, bool from_host)
 #endif
 	default:
+#ifdef ENABLE_HOST_FIREWALL
+		ret = DROP_UNKNOWN_L3;
```
Should the default host firewall drop code be different?

pchaigno replied:
If I'm not mistaken, if `DROP_NOTIFY` is defined, the `EVENT_SOURCE` (i.e., `HOST_EP_ID` for the host endpoint) will be logged when the drop is logged. I don't know how often `DROP_NOTIFY` is set, but if that's considered to be the default, it might not be worth introducing new drop codes? On the other hand, the only downside of adding new drop codes is that it takes a couple new slots and the limit on the max. number of drop codes we can have (tied to metrics map's size) is still far.

Unrelated: Your comment made me realize that we're missing a call to `send_drop_notify_error()` here.

joestringer replied:
> if DROP_NOTIFY is defined, the EVENT_SOURCE (i.e., HOST_EP_ID for the host endpoint) will be logged when the drop is logged.

My concern was if we end up reporting to the user "we're dropping this traffic because we don't support this protocol", but we really mean "the host firewall dropped the traffic and the firewall requires ipv4/ipv6". That said this is not likely to hit particularly often and it's likely unique at this `EVENT_SOURCE` anyway so we can figure out the reason here in retrospect. I agree that it's probably not necessary to make a distinction at this point.

FWIW, `DROP_NOTIFY `is typically defined to true. This is how we get notifications when packets are dropped.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -800,9 +1061,47 @@ int to_host(struct __ctx_buff *ctx)
+	case bpf_htons(ETH_P_IP):
+		ret = ipv4_host_policy_ingress(ctx, &srcID);
+		break;
+# endif
```
The "redirect to proxy" calls above do not steal the packet away to redirect to the proxy today, so I would expect such traffic to fall through to here and that this would break pod L7 policy if you enable host firewall policies. Presumably we should carefully detect that case and bypass the host firewall?

pchaigno replied:
If I understand correctly, these packets (with `MARK_MAGIC_TO_PROXY`) are on their way to the proxy from an endpoint, and you propose we whitelist that path (`ep -> proxy`)? Wouldn't we also need to whitelist the return path (`proxy -> ep`)?

I'm really unfamiliar with the L7 proxy stuff, but couldn't we express this whitelisting as a host policy (ingress to a specific port maybe)? Then we could have it as a sort of default host policy.

joestringer replied:
> If I understand correctly, these packets (with MARK_MAGIC_TO_PROXY) are on their way to the proxy from an endpoint, and you propose we whitelist that path (ep -> proxy)? Wouldn't we also need to whitelist the return path (proxy -> ep)?

:+1: 

Forward path here should be straightforward. The knowledge we need is just a few lines above this point.

Traffic from the proxy we can identify using the packet mark (Proxy will always configure the SO_MARK on the socket; this is also how we inherit the original source identity for the flow). Looking at the "inherit identity from stack" functions should provide you the context here.

> I'm really unfamiliar with the L7 proxy stuff, but couldn't we express this whitelisting as a host policy (ingress to a specific port maybe)? Then we could have it as a sort of default host policy.

Plausibly but a couple of caveats:
(1) The proxy port is not in the packet itself; TProxy allows us to redirect traffic without modifying the packet tuple.
(2) Adding the port to the proxy would presumably allow any peer to send traffic towards the proxy. It wouldn't just be traffic to a pod that is subject to L7 policy. So the proxy allow could potentially be broader than necessary.

pchaigno replied:
I pushed a fix for this in the new commit `bpf: Bypass host policies for packets to/from proxy`. I haven't validated it yet (besides `-C bpf build_all`), but pushed anyway to run CI tests. I'll validate manually in parallel.

pchaigno replied:
As discussed offline and commented by Thomas below, that commit to bypass the host policies for packets to/from proxy isn't needed:
- Packets redirected to the L7 proxy will have the IPs of the pods and therefore will already bypass the host policies.
- Packets from the proxy to the server (establishment of TCP handshake once handshake with client pod is done) will have the identity of the client pod embedded in their mark and that will be resolved correctly by the current bpf_host code, thereby bypassing the host policies again.
- Packets on the return path (e.g., TCP SYN+ACK from server pod to proxy) will match a CT entry inserted by the forward path packet.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -333,14 +363,178 @@ resolve_srcid_ipv4(struct __ctx_buff *ctx, __u32 srcid_from_proxy,
+	tuple.daddr = ip4->daddr;
+	tuple.saddr = ip4->saddr;
+	l4_off = l3_off + ipv4_hdrlen(ip4);
+	ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_EGRESS,
```
Any thoughts on how to avoid an overly chatty pod consuming all CT entries and preventing host connectivity?

pchaigno replied:
Apart from fixing per-endpoint CT maps and switching the host endpoint to always use its own CT map, not really :-/

joestringer replied:
OK. I would strongly encourage figuring out how to resolve this before we mark the feature as "stable".

joestringer commented on pkg/policy/api/rule_validation.go:
```diff
@@ -46,12 +46,23 @@ func (r Rule) Sanitize() error {
+	if r.NodeSelector.LabelSelector != nil {
+		if err := r.NodeSelector.sanitize(); err != nil {
+			return err
+		}
```
(Not this line; but IIRC this file)

I think we should also reject policies with both `NodeSelector` and L7 policy since that is currently unsupported?

(Could be followup change if you like)

pchaigno replied:
Done and added unit tests for this.

joestringer reviewed (changes requested):
I made a first pass with mostly high-level feedback discussion points below, although I couldn't pick many nits anyway as it overall looks really clean!

I didn't look closely at the last 3 patches (ipv6 and the complexity/lb patches). Several of my points below are appropriate for separate followup later.

aanm commented on examples/policies/l3/host/host-policy.json:
```diff
@@ -0,0 +1,41 @@
+[
+  {
```
can you add an example of a k8s CNP in yaml as well?

pchaigno replied:
It looks like I do need a couple changes to enable `nodeSelector` in CNPs :-/
```
$ cat examples/policies/l3/host/host-policy.yaml
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: "allow-dns-on-udp-53-only"
spec:
  nodeSelector:
    matchLabels:
      {}
  ingress:
  - toPorts:
    - ports:
      - port: "6443"
        protocol: TCP
      - port: "22"
        protocol: TCP
  egress:
  - toPorts:
    - ports:
      - port: "4240"
        protocol: TCP
      - port: "8080"
        protocol: TCP
  - toCIDR:
    - "8.8.0.0/16"
    toPorts:
    - ports:
      - port: "53"
        protocol: UDP
$ kubectl create -f examples/policies/l3/host/host-policy.yaml 
The CiliumNetworkPolicy "allow-dns-on-udp-53-only" is invalid: spec.endpointSelector: Required value
```
I haven't looked at how much changes are required, but maybe best to address in a follow up?

joestringer replied:
I don't mind getting the datapath bits in here first but obviously being able to install these policies with CNP is important before v1.8 full release. I don't expect this to be a big change.

pchaigno replied:
I took a second look at this and found a couple things to fix, but I'm still having that same error message. I think I'll need to move this to follow ups :-/

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -34,8 +37,18 @@ import (
+	// +optional
+	EndpointSelector EndpointSelector `json:"endpointSelector,omitempty"`
+
+	// EndpointSelector selects all endpoints which should be subject to
```
it should start with `NodeSelector selects...`

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -70,6 +83,63 @@ func NewRule() *Rule {
 
+// MarshalJSON returns the JSON encoding of Rule r. We need to overwrite it to
+// enforce omitempty on the EndpointSelector nested structures.
+func (r *Rule) MarshalJSON() ([]byte, error) {
```
Discussed offline

pchaigno replied:
Added a note to rework this in follow ups because it's less trivial than we thought. Using `reflect`, I might have a solution that doesn't require as many code changes and is easier on maintainability, but it still needs some work.

christarazi replied:
I just noticed that this PR introduced custom JSON marshaling for `Rule`. This has some implications for https://github.com/cilium/cilium/pull/11607. Specifically, one of the reasons I forked `controller-tools` is to remove a constraint which says: any type implementing custom JSON marshaling will have its validation schema replaced with `type: Any`. In the upstream, https://github.com/kubernetes-sigs/controller-tools/pull/427 is responsible for this. In the fork, I've reverted that support. 

My question is why do we need this? I [asked on K8s Slack](https://kubernetes.slack.com/archives/CAR30FCJZ/p1596152053329900?thread_ts=1596146930.326300&cid=CAR30FCJZ) about opting out of this feature in controller-tools. It's possible that we can keep it with a knob in controller-tools, or we solve the problem another way.

pchaigno replied:
We need this to properly implement the `omitempty` tags of `NodeSelector` and `EndpointSelector`. Because they are not pointers, without the custom marshal, marshalling then unmarshalling would create a new field where none existed. That is, when marshalling it would check `r.EndpointSelector` (always `!= nil`) instead of `r.EndpointSelector.LabelSelector` and thus always create the corresponding JSON entry.

You can easily reproduce this by removing the custom marshalling and running unit tests on the package. One of the tests checks `json.Unmarshal(json.Marhsal(rule)) == rule`.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1564,6 +1565,42 @@ func initClockSourceOption() {
+	//                    (ENABLE_EXTERNAL_IP && !BPF_HAVE_NETNS_COOKIE))
+	// We can't enable both egress LB and host firewall on kernels <4.14 due to
+	// the verifier complexity limit, at 96k instructions.
+	netnsCookieSupport := false
```
var netnsCookieSupport bool

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1564,6 +1565,42 @@ func initClockSourceOption() {
+	// the verifier complexity limit, at 96k instructions.
+	netnsCookieSupport := false
+	pm := probes.NewProbeManager()
+	if h1, h2 := pm.GetHelpers("cgroup_sock_addr"), pm.GetHelpers("cgroup_sock"); h1 != nil && h2 != nil {
```
You can leave the code as it is but just a FYI you can access nil maps. https://play.golang.org/p/t__FGNkDvCx

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1564,6 +1565,42 @@ func initClockSourceOption() {
+		// We can probe the verifier's complexity limit (128k here) indirectly
+		// based on what kernel helpers are available since both were changed
+		// in the same release, v4.14.
+		found := false
```
var found bool

aanm reviewed (changes requested):
Only small nits, overall LGTM. I only reviewed the control plane sections

qmonnet reviewed (approved):
The set looks good to me, nice work!

pchaigno committed:
daemon: Introduce enable-host-firewall option

BPF-based egress load balancing isn't supported from the host namespace,
so the host policies will fail (due to wrong destination endpoint) if
that was enabled. Most users likely use bpf_sock (newer kernels) or
kube-proxy for egress load balancing so we just disable BPF-based egress
load balancing with a warning if the host firewall is enabled.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
pkg/policy: Host network policies

This commit adds network policies for nodes. Node (or host) network
policies are loaded with a new nodeSelector field. The nodeSelector and
endpointSelector fields cannot co-exist for now. Rules with a nodeSelector
field apply only to the special endpoint representing the host.

Network policies for the host cannot contain L7 policies yet as this
isn't supported in the datapath.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Enforce host policies for IPv6

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: localID as parameter of policy logic functions

bpf_lxc programs have their localID hardcoded as SECLABEL, but that is
not the case for the bpf_host program in the case of the host firewall.
This commit prepares the policy_can_{egress,access_ingress} functions to
take the localID as an argument.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Hints to verifier's state pruning

Introduce pruning checkpoints to help the verifier's state pruning on
older kernels. All program sections now load with less than 128k
processed instructions on v4.9.212 and less than 65k on v4.14.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Enforce host policies for IPv4

Host policies are only enforced when --enable-host-firewall is set. Once
enforced, they implement a default deny.

Signed-off-by: Paul Chaignon <paul@cilium.io>

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1129,6 +1133,17 @@ func initEnv(cmd *cobra.Command) {
+	if option.Config.EnableHostFirewall && len(option.Config.Devices) == 0 {
+		device, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
+		if err != nil {
+			msg := "Host firewall's external facing device could not be determined. Use --device to specify."
```
`"Use --%s to specifiy", option.Device)`

aanm reviewed (approved):
Only one small nit, can be addressed as a follow up

tgraf commented on bpf/bpf_host.c:
```diff
@@ -370,8 +566,6 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx, bool from_host)
 	tuple.nexthdr = ip4->protocol;
 
 	if (from_host) {
-		int ret;
```
Doesn't the:
```
		return CTX_ACT_OK;
```

above in the `!from_host` branch bypass the host firewall?

tgraf commented on bpf/bpf_host.c:
```diff
@@ -428,10 +438,13 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, bool from_host)
-	else
+	if (from_host) {
+		/* Bypass host policies if on return path from proxy. */
+		if (!is_from_proxy(ctx))
```
I'm surprised that this is needed. Shouldn't `secctx` be the actual security identity other than `HOST_ID` and thus lead to an allow?

pchaigno replied:
That is correct and the commit that added bypassing of the host policies for the L7 proxy actually isn't needed. See https://github.com/cilium/cilium/pull/11507#discussion_r426893006 for details. I've removed it.

brb commented on bpf/bpf_host.c:
```diff
@@ -188,6 +188,34 @@ ipv6_host_policy_egress(struct __ctx_buff *ctx, __u32 srcID)
 		return hdrlen;
 	l4_off = l3_off + hdrlen;
+
+#ifdef ENABLE_SERVICES
```
Can we drop the service handling? Asking, as this implementation is not enough to properly handle a reverse path (e.g. reply coming from a pod running in the host netns (e.g. kube-apiserver) won't hit bpf_host.o, also a reply from a local service endpoint is not rev-DNAT xlated).

It's safe enough to assume that a user will either be running with bpf_sock or have kube-proxy (in both cases the host fw should see a real endpoint IP addr instead of VIP).

pchaigno replied:
> Can we drop the service handling? Asking, as this implementation is not enough to properly handle a reverse path (e.g. reply coming from a pod running in the host netns (e.g. kube-apiserver) won't hit bpf_host.o, also a reply from a local service endpoint is not rev-DNAT xlated).

Sure, less complexity issues. I'll remove commits `bpf: Support LB on host egress` and `daemon: Disable egress LB if EnableHostFirewall on older kernels` on the next iteration in the evening. I'll also disable service handling in BPF with a warning if the user enabled the host firewall.

pchaigno replied:
I removed those commits and, in commit `daemon: Introduce enable-host-firewall option`, I disabled BPF-based LB handling with a warning when the host firewall is enabled.

gandro commented on daemon/cmd/daemon_main.go:
```diff
@@ -1132,6 +1136,17 @@ func initEnv(cmd *cobra.Command) {
+	if option.Config.EnableHostFirewall && len(option.Config.Devices) == 0 {
+		device, err := linuxdatapath.NodeDeviceNameWithDefaultRoute()
+		if err != nil {
+			msg := "Host firewall's external facing device could not be determined. Use --%s to specify."
```
On local testing, this `%s` did not format as expected. I think you have to use `Fatalf` below.

pchaigno replied:
Ah, I thought Go would complain if `Fatal()` was incorrect and didn't check. I've sent a fix.

Thanks a lot for trying it out @gandro!

gandro reviewed (commented):
Nice! I did a couple of smoke tests with Hubble, and everything worked as expected (endpoint ID, security identity and labels for host traffic was annotated correctly, drop/trace events were observed as expected with the example host policy installed).

Of course for Hubble-Relay (which opens a TCP port on the host) we will need to document that people need to whitelist that port if they use host firewall, but this is conceptually very similar to the health checks, so no concern there.

In terms of code, only one very minor issue with the output (broken formatting of error message), see comment.
